src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlConnectionDetails.cs

```diff
@@ -0,0 +1,126 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Collections.Concurrent;
+using System.Text.RegularExpressions;
+
+namespace OpenTelemetry.Instrumentation.SqlClient.Implementation;
+
+internal sealed class SqlConnectionDetails
+{
+    /*
+     * Match...
+     *  protocol[ ]:[ ]serverName
+     *  serverName
+     *  serverName[ ]\[ ]instanceName
+     *  serverName[ ],[ ]port
+     *  serverName[ ]\[ ]instanceName[ ],[ ]port
+     *
+     * [ ] can be any number of white-space, SQL allows it for some reason.
+     *
+     * Optional "protocol" can be "tcp", "lpc" (shared memory), or "np" (named pipes). See:
+     *  https://docs.microsoft.com/troubleshoot/sql/connect/use-server-name-parameter-connection-string, and
+     *  https://docs.microsoft.com/dotnet/api/system.data.sqlclient.sqlconnection.connectionstring?view=dotnet-plat-ext-5.0
+     *
+     * In case of named pipes the Data Source string can take form of:
+     *  np:serverName\instanceName, or
+     *  np:\\serverName\pipe\pipeName, or
+     *  np:\\serverName\pipe\MSSQL$instanceName\pipeName - in this case a separate regex (see NamedPipeRegex below)
+     *  is used to extract instanceName
+     */
+    private static readonly Regex DataSourceRegex = new("^(.*\\s*:\\s*\\\\{0,2})?(.*?)\\s*(?:[\\\\,]|$)\\s*(.*?)\\s*(?:,|$)\\s*(.*)$", RegexOptions.Compiled);
+
+    /// <summary>
+    /// In a Data Source string like "np:\\serverName\pipe\MSSQL$instanceName\pipeName" match the
+    /// "pipe\MSSQL$instanceName" segment to extract instanceName if it is available.
+    /// </summary>
+    /// <see>
+    /// <a href="https://docs.microsoft.com/previous-versions/sql/sql-server-2016/ms189307(v=sql.130)"/>
+    /// </see>
+    private static readonly Regex NamedPipeRegex = new("pipe\\\\MSSQL\\$(.*?)\\\\", RegexOptions.Compiled);
```

Comment:
There is a RegEx source generator available now. Doesn't have to be part of this PR, but we should _probably_ move to that on targets where it is available. Reason being is I assume it gives best perf across the board but especially when AoT is in play and compilation is not available (assuming RegEx falls back to interpreter when dynamic code isn't available).

Some reasons listed here why it should be used:
https://learn.microsoft.com/dotnet/standard/base-types/regular-expression-source-generators#when-to-use-it

Reply:
#2234